### PR TITLE
fix(optimizer)!: Refactor struct star expansion in BQ

### PR DESCRIPTION
**Pull request by VaggelisD**
Let's assume we want to expand the following struct:

```
WITH
   tbl1 AS (SELECT STRUCT(1 AS fld1, Struct(3 AS fld2, Struct(4 AS fld3) AS lvl2) AS lvl1) AS col)
SELECT col.* FROM tbl1;
```

The recently merged PR would essentially flatten the struct by exploding the fields through a DFS traversal, i.e it would produce:

```
WITH
   tbl1 AS (...) # redacted CTE, same as initial query
SELECT tbl1.col.fld1 AS _f_0, 
       tbl1.col.lvl1.fld2 AS _f_2,
       tbl1.col.lvl1.lvl2.fld3 AS _f_3,
FROM tbl1
```

Even though the results are similar to what would be produced from the initial query, internally BigQuery only has to expand [the _top level_ expression](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#select_expression_star) and the rest of the "flattening" is performed implicitly; This is semantically equivalent to:

```
WITH
   tbl1 AS (...)
SELECT tbl1.col.fld1 AS fld1, 
       tbl1.col.lvl1 AS lvl1 
FROM tbl1
```

By changing our algorithm to perform the latter we completely remove the DFS struct traversal. Not only that, but it turns out that our `_f_` alias generation can break queries from being qualified; The following query can showcase this behavior:

```
WITH
  tbl1 AS (SELECT STRUCT(1 AS fld1, Struct(3 AS fld2, Struct(4 AS fld3) AS lvl2) AS lvl1) AS col),
  tbl2 AS (SELECT col.* FROM tbl1)
SELECT lvl1 FROM tbl2
```

In main, due to our custom `_f_` aliasing, `qualify` fails with:
```
sqlglot.errors.OptimizeError: Column '"lvl1"' could not be resolved
```

However, in this PR, `qualify` succeeds (the names are now preserved instead of generating a custom alias) and results to:
```
WITH 
`tbl1` AS (SELECT STRUCT(1 AS `fld1`, Struct(3 AS `fld2`, Struct(4 AS `fld3`) AS `lvl2`) AS `lvl1`) AS `col`), 
`tbl2` AS (SELECT `tbl1`.`col`.`fld1` AS `fld1`, `tbl1`.`col`.`lvl1` AS `lvl1` FROM `tbl1` AS `tbl1`) 
SELECT 
  `tbl2`.`lvl1` AS `lvl1` FROM `tbl2` AS `tbl2`
```

### CODE REVIEW

**Comment by georgesittas** [comment]
Chiming in here to provide some more context around this:

> Even though the results are similar to what would be produced from the initial query, internally BigQuery only has to expand [the top level expression](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#select_expression_star) and the rest of the "flattening" is **performed implicitly**

TL;DR - the previous implementation misinterpreted BQ's struct expansion semantics, most likely because of how they display the column names in their editor:

```sql
-- BQ displays a single row with a single column named s.id
with t as (select struct(struct('foo' as id) as s) as col) select col.* from t
```

File: sqlglot/optimizer/qualify_columns.py [hunk 1]
```diff
@@ -441,46 +438,22 @@ def _expand_struct_stars(
...
 
-    # Each nested field caches its path in exp.Identifier parts,
-    # e.g. tbl.f0.f1.item -> [exp.Identifier(tbl), ..., exp.Identifier(item)]
-    field_path: t.Dict[int, t.List[exp.Identifier]] = {id(starting_struct): []}
-
-    # The keys to non-struct field entries in `nested_fields`
-    field_keys = []
-
-    stack = [(starting_struct, id(starting_struct))]
+    name_set = set()
```
Comment (georgesittas) [inline sqlglot/optimizer/qualify_columns.py:441]: Rename this to `taken_names` or something similar.

File: sqlglot/optimizer/qualify_columns.py [hunk 2]
```diff
@@ -441,46 +438,22 @@ def _expand_struct_stars(
...
-    for field_key in field_keys:
-        # The fully qualified path of each field is created by:
-        # - The path until the starting struct, if any (dot_parts)
-        # - The path from the starting struct until the nested field (field_path[field_key])
-        root, *parts = [
-            part.copy() for part in itertools.chain(dot_parts[:-1], field_path[field_key])
-        ]
+        this = field.this.copy()
+        root, *parts = [part.copy() for part in itertools.chain(dot_parts, [this])]
         new_column = exp.column(t.cast(exp.Identifier, root), table=dot_column.table, fields=parts)
```
Comment (georgesittas) [inline sqlglot/optimizer/qualify_columns.py:455]: I think if the `dot_column`'s table contained quotes, you'd drop them in `table=dot_column.table` because under the hood it simply invokes `text("table")`. You need to do `dot_column.args.get("table")`.

Let's 1) ensure this _does_ indeed drop the quotes in an example where the table name is quoted, 2) fix it and 3) add a test to avoid similar regressions.

Please also check for similar property method invocations to ensure we don't silently drop quotes.

**Review by georgesittas — APPROVED** [review]
Thanks for fixing this
